File: Common/Closable.vb
Namespace Common

    Public Class Closable
        Implements IDisposable

        Public Sub New(ByVal CloseOperation As Action)
            Me.CloseOperation = CloseOperation
        End Sub

        Public Sub Close()
            Me.CloseOperation.Invoke()
        End Sub

        Private Sub Dispose() Implements IDisposable.Dispose
            Me.Close()
        End Sub

        Private ReadOnly CloseOperation As Action

    End Class

End Namespace

File: Common/Closable.cs
using System;

namespace Ks.Common
{
    public class Closable : IDisposable
    {
        public Closable(Action CloseOperation)
        {
            this.CloseOperation = CloseOperation;
        }

        public void Close()
        {
            this.CloseOperation.Invoke();
        }

        void IDisposable.Dispose()
        {
            this.Close();
        }

        private readonly Action CloseOperation;
    }
}
